### PR TITLE
feat(content-sidebar): added support for active comment reply

### PR DESCRIPTION
**Pull request by kkuliczkowski-box**
Added support for handling links with active item that is a reply. This is to support mentioning users directly in a reply. The link that user will receive in notification email will have only type and id of the feed item. This change covers that scenario by doing additional request for that replie's details (in order to get id of its parent). Logic can be seen in the diagram:

handling-active-comment-reply

**Please note that this is a temporary solution done based on time constraints presented by Product. Final solution would be to change the link in notification email to also include `parentId`, thus making this additional request mentioned above not needed.**

### CODE REVIEW

File: src/elements/content-sidebar/activity-feed/activity-feed/ActivityFeed.js [hunk 1]
```diff
@@ -110,23 +125,38 @@ class ActivityFeed extends React.Component<Props, State> {
             currentUser: prevCurrentUser,
             feedItems: prevFeedItems,
         } = prevProps;
-        const { feedItems: currFeedItems, activeFeedEntryId } = this.props;
-        const { isInputOpen: prevIsInputOpen } = prevState;
-        const { isInputOpen: currIsInputOpen } = this.state;
+        const { feedItems: currFeedItems, activeFeedEntryId, onShowReplies = noop } = this.props;
+        const { activeEntry, isInputOpen: prevIsInputOpen } = prevState;
+        const { activeEntry: prevActiveEntry, isInputOpen: currIsInputOpen } = this.state;
```
Comment (jstoffan) [inline src/elements/content-sidebar/activity-feed/activity-feed/ActivityFeed.js:130]: Is this inverted? It looks like this should be `currActiveEntry` and the line above should be `prevActiveEntry`.

File: src/elements/content-sidebar/activity-feed/activity-feed/ActivityFeed.js [hunk 2]
```diff
@@ -110,23 +125,38 @@ class ActivityFeed extends React.Component<Props, State> {
             currentUser: prevCurrentUser,
             feedItems: prevFeedItems,
         } = prevProps;
-        const { feedItems: currFeedItems, activeFeedEntryId } = this.props;
-        const { isInputOpen: prevIsInputOpen } = prevState;
-        const { isInputOpen: currIsInputOpen } = this.state;
+        const { feedItems: currFeedItems, activeFeedEntryId, onShowReplies = noop } = this.props;
```
Comment (jstoffan) [inline src/elements/content-sidebar/activity-feed/activity-feed/ActivityFeed.js:128]: Let's use defaultProps for the default `noop` value here.

File: src/elements/content-sidebar/activity-feed/activity-feed/ActivityFeed.js [hunk 3]
```diff
@@ -110,23 +125,38 @@ class ActivityFeed extends React.Component<Props, State> {
...
         const didLoadFeedItems = prevFeedItems === undefined && currFeedItems !== undefined;
         const hasInputOpened = currIsInputOpen !== prevIsInputOpen;
         const hasActiveFeedEntryIdChanged = activeFeedEntryId !== prevActiveFeedEntryId;
+        const hasActiveEntryChanged = activeEntry?.id !== prevActiveEntry?.id;
 
         if ((hasLoaded || hasMoreItems || didLoadFeedItems || hasInputOpened) && activeFeedEntryId === undefined) {
             this.resetFeedScroll();
         }
 
-        if (didLoadFeedItems || hasActiveFeedEntryIdChanged) {
```
Comment (jstoffan) [inline src/elements/content-sidebar/activity-feed/activity-feed/ActivityFeed.js:127]: Do we still need to check `didLoadFeedItems`? If not, won't this fail in most circumstances?

File: src/elements/content-sidebar/activity-feed/activity-feed/ActivityFeed.js [hunk 4]
```diff
@@ -88,12 +99,16 @@ type Props = {
...
 
 type State = {
+    activeEntry?: FeedItem | null,
     isInputOpen: boolean,
+    isReplyDataLoading: boolean,
 };
 
 class ActivityFeed extends React.Component<Props, State> {
     state = {
+        activeEntry: undefined,
```
Comment (jstoffan) [inline src/elements/content-sidebar/activity-feed/activity-feed/ActivityFeed.js:109]: We should avoid storing derived state from props in this way, since it can lead to synchronization issues. Let's retrieve the entry as needed using the value from `props.activeFeedEntryId`.

File: src/elements/content-sidebar/activity-feed/activity-feed/ActivityFeed.js [hunk 5]
```diff
@@ -160,6 +190,61 @@ class ActivityFeed extends React.Component<Props, State> {
         return feedItems.length === 0 || (feedItems.length === 1 && feedItems[0].type === ItemTypes.fileVersion);
     };
 
+    getActiveEntry = (callback: (activeEntry: FeedItem | null, repliesNeedToBeLoaded?: boolean) => void) => {
```
Comment (jstoffan) [inline src/elements/content-sidebar/activity-feed/activity-feed/ActivityFeed.js:193]: Can we use a promise or async/await instead of introducing callbacks?

File: src/elements/content-sidebar/activity-feed/activity-feed/ActivityFeed.js [hunk 6]
```diff
@@ -160,6 +190,61 @@ class ActivityFeed extends React.Component<Props, State> {
...
+        if (!activeFeedEntryId || !Array.isArray(feedItems)) {
+            return callback(null);
+        }
+
+        const firstLevelItem = feedItems.find(
+            ({ id, type }) => id === activeFeedEntryId && type === activeFeedEntryType,
+        );
+        // If the active entry is a first level Feed item, return it
+        if (firstLevelItem) {
+            return callback(firstLevelItem);
```
Comment (jstoffan) [inline src/elements/content-sidebar/activity-feed/activity-feed/ActivityFeed.js:205]: We generally want to avoid multiple returns, as it makes a function very difficult to debug and reason about. This function is also _very_ complicated and probably needs to be decomposed.

File: src/elements/content-sidebar/activity-feed/activity-feed/ActivityFeed.js [hunk 7]
```diff
@@ -160,6 +190,61 @@ class ActivityFeed extends React.Component<Props, State> {
...
+            return callback(null);
+        }
+
+        const firstLevelItemWithActiveReply = feedItems.find(item => {
+            if (
+                (item.type !== FEED_ITEM_TYPE_ANNOTATION && item.type !== FEED_ITEM_TYPE_COMMENT) ||
+                !item.replies ||
+                !Array.isArray(item.replies)
+            ) {
+                return false;
```
Comment (jstoffan) [inline src/elements/content-sidebar/activity-feed/activity-feed/ActivityFeed.js:219]: Should this return undefined?

File: src/api/ThreadedComments.js [hunk 8]
```diff
@@ -228,6 +228,45 @@ class ThreadedComments extends MarkerBasedApi {
...
+        permissions,
+        successCallback,
+    }: {
+        commentId: string,
+        errorCallback: (e: ElementsXhrError, code: string) => void,
+        fileId: string,
+        permissions: BoxItemPermission,
+        successCallback: (comment: Comment) => void,
+    }): void {
+        this.errorCode = ERROR_CODE_FETCH_COMMENTS;
```
Comment (jstoffan) [inline src/api/ThreadedComments.js:254]: Should this use a new error code?

File: src/elements/content-sidebar/activity-feed/activity-feed/ActivityFeed.js [hunk 9]
```diff
@@ -160,6 +190,61 @@ class ActivityFeed extends React.Component<Props, State> {
...
+            }
+            return item.replies.some(({ id }) => id === activeFeedEntryId);
+        });
+        // If the active entry is within present replies of any first level Feed items (if it's a comment or annotation),
+        // return that first level Feed item
+        if (firstLevelItemWithActiveReply) {
+            return callback(firstLevelItemWithActiveReply);
+        }
+
+        if (!getComment) {
```
Comment (jstoffan) [inline src/elements/content-sidebar/activity-feed/activity-feed/ActivityFeed.js:229]: Is this possible with the current implementation? It looks like the method is always defined.

File: src/elements/content-sidebar/activity-feed/activity-feed/ActivityFeed.js [hunk 10]
```diff
@@ -43,6 +49,11 @@ type Props = {
     file: BoxItem,
     getApproverWithQuery?: Function,
     getAvatarUrl: GetAvatarUrlCallback,
+    getComment?: (
```
Comment (jstoffan) [inline src/elements/content-sidebar/activity-feed/activity-feed/ActivityFeed.js:52]: Should we handle this case in `ActivitySidebar.fetchFeedItems` rather than inside ActivityFeed?

File: src/elements/content-sidebar/activity-feed/activity-feed/ActivityFeed.js [hunk 11]
```diff
@@ -298,7 +378,7 @@ class ActivityFeed extends React.Component<Props, State> {
                     }}
                     className="bcs-activity-feed-items-container"
                 >
-                    {isLoading && (
+                    {(isLoading || isReplyDataLoading) && (
```
Comment (jstoffan) [inline src/elements/content-sidebar/activity-feed/activity-feed/ActivityFeed.js:381]: The combination of loading state here indicates again that this should probably occur when the feed items are being fetched rather than separately. If not, though, should this move into the `hasLoaded` method?

File: src/elements/content-sidebar/activity-feed/activity-feed/ActivityFeed.js [hunk 12]
```diff
@@ -313,8 +393,8 @@ class ActivityFeed extends React.Component<Props, State> {
                     {!isEmpty && !isLoading && (
                         <ActiveState
                             {...activityFeedError}
-                            activeFeedEntryId={activeFeedEntryId}
-                            activeFeedEntryType={activeFeedEntryType}
+                            activeFeedEntryId={activeEntry?.id}
```
Comment (jstoffan) [inline src/elements/content-sidebar/activity-feed/activity-feed/ActivityFeed.js:396]: Let's revert these changes given that we want to avoid derived state.

**Review by jstoffan — CHANGES_REQUESTED** [review]

File: src/elements/content-sidebar/activity-feed/activity-feed/ActivityFeed.js [hunk 13]
```diff
@@ -160,6 +190,61 @@ class ActivityFeed extends React.Component<Props, State> {
...
+            activeFeedEntryId,
+            ({ parent }) => {
+                this.setState({ isReplyDataLoading: false });
+                const parentItem = feedItems.find(({ id }) => id === parent?.id) || null;
+                callback(parentItem, true);
+            },
+            () => {
+                this.setState({ isReplyDataLoading: false });
+                callback(null);
+            },
```
Comment (tjuanitas) [inline src/elements/content-sidebar/activity-feed/activity-feed/ActivityFeed.js:244]: where is `getComment` initially from passed from? i'm curious how the order of these args are determined.

File: src/elements/content-sidebar/activity-feed/activity-feed/ActivityFeed.js (hunk 12 above)
Comment (tjuanitas) [inline src/elements/content-sidebar/activity-feed/activity-feed/ActivityFeed.js:396]: is there a reason we call the state `activeEntry` instead of `activeFeedEntry`?

**Comment by kkuliczkowski-box** [comment]
I've changed the approach and followed @jstoffan's advice and moved the whole logic of fetching the reply data to `ActivitySidebar`. Thanks to that we don't need this additional "loading" step and the reply data is being loaded before the feed items are displayed. Because all of the changes done in `ActivityFeed` have been reverted, I haven't responded to most of the comments.

I'm struggling with Flow check at the moment and will continue working on it.

File: src/elements/content-sidebar/activity-feed/activity-feed/ActivityFeed.js (hunk 7 above)
Comment (kkuliczkowski-box) [reply src/elements/content-sidebar/activity-feed/activity-feed/ActivityFeed.js]: I think It can return `false` as this is the return statement of the `find` callback function. I've copied that piece of logic into `ActivitySidebar`.

File: src/elements/content-sidebar/ActivitySidebar.js [hunk 14]
```diff
@@ -152,20 +158,31 @@ class ActivitySidebar extends React.PureComponent<Props, State> {
...
-        const { logger } = this.props;
+        const { activeFeedEntryId, activeFeedEntryType, logger } = this.props;
 
         mark(MARK_NAME_DATA_LOADING);
 
         logger.onReadyMetric({
             endMarkName: MARK_NAME_JS_READY,
         });
-        this.state = {};
+        this.state = { activeFeedEntryId, activeFeedEntryType, approverSelectorContacts: [] };
```
Comment (jstoffan) [inline src/elements/content-sidebar/ActivitySidebar.js:168]: We should almost always avoid duplicating our props or any prop-derived value in state, as it can easily lead to synchronization issues. Let's update the logic to use the values directly from props.

File: src/elements/content-sidebar/activity-feed/activity-feed/ActivityFeed.js (hunk 10 above)
Comment (kkuliczkowski-box) [reply src/elements/content-sidebar/activity-feed/activity-feed/ActivityFeed.js]: That makes a lot of sense. I've moved the logic for fetching reply data to `ActivitySidebar` but I've placed it in `fetchFeedItemsSuccessCallback` as I didn't want to change the original purpose of `fetchFeedItems. It seemed better to put it in `fetchFeedItemsSuccessCallback` as I'm taking actions based on the success of the `fetchFeedItems`.

File: src/api/Feed.js [hunk 15]
```diff
@@ -462,6 +462,51 @@ class Feed extends Base {
...
+
+    /**
+     * Callback for successful fetch of a comment
+     *
+     * @param {Function} resolve - resolve function
+     * @param {Function} successCallback - success callback
+     * @param {Comment} comment - comment data
+     * @return {void}
+     */
+    fetchThreadedCommentSuccessCallback = (resolve: Function, successCallback: Function, comment: Comment): void => {
```
[the remark this answers is not in the document]
Comment (kkuliczkowski-box) [reply src/api/Feed.js:505]: I needed to keep `comment` as the last param because I'm binding this method in `fetchThreadedComment` method.

File: src/elements/content-sidebar/ActivitySidebar.js [hunk 16]
```diff
@@ -717,7 +738,34 @@ class ActivitySidebar extends React.PureComponent<Props, State> {
...
+            activeFeedEntryType !== FEED_ITEM_TYPE_COMMENT
+        ) {
+            this.setState({ ...baseChanges });
+            return;
+        }
+
+        this.getActiveFeedEntryData(feedItems)
+            .then(({ id, type }) => {
+                this.setState({
+                    ...baseChanges,
```
Comment (jstoffan) [inline src/elements/content-sidebar/ActivitySidebar.js:761]: I'm a bit concerned about race conditions here. What happens if `fetchFeedItems` is called again while this promise is still in progress? Would we wind up with stale feed items? It seems like we'd want to handle this conditional promise chain prior to the success callback being invoked?

File: src/elements/content-sidebar/ActivitySidebar.js [hunk 17]
```diff
@@ -717,7 +738,34 @@ class ActivitySidebar extends React.PureComponent<Props, State> {
             );
         }
 
-        this.setState({ feedItems, activityFeedError: undefined });
+        const baseChanges = { feedItems, activityFeedError: undefined };
+
+        // If activeFeedEntryId is not set or
```
Comment (jstoffan) [inline src/elements/content-sidebar/ActivitySidebar.js:743]: This comment is largely describing the logic below. We should use comments to explain "why" something is written if it's not obvious in the code.

File: src/elements/content-sidebar/ActivitySidebar.js [hunk 18]
```diff
@@ -827,6 +875,149 @@ class ActivitySidebar extends React.PureComponent<Props, State> {
...
+                if (
+                    type === FEED_ITEM_TYPE_ANNOTATION ||
+                    type === FEED_ITEM_TYPE_COMMENT ||
+                    type === FEED_ITEM_TYPE_TASK
+                ) {
+                    resolve({ id, type });
+                    return;
+                }
+                resolve({ id: activeFeedEntryId, type: activeFeedEntryType });
+                return;
```
Comment (jstoffan) [inline src/elements/content-sidebar/ActivitySidebar.js:958]: We should avoid multiple returns, as it makes logic very difficult to debug and reason about. We're doing quite a bit of duplicate work in this function and it would be great to refactor it for simplicity.

File: src/elements/content-sidebar/ActivitySidebar.js [hunk 19]
```diff
@@ -827,6 +875,149 @@ class ActivitySidebar extends React.PureComponent<Props, State> {
...
+
+    /**
+     * Fetches comment data
+     *
+     * @param {string} id - id of the comment
+     * @param {Function} onSuccess - the success callback
+     * @param {ErrorCallback} onError -the error callback
+     * @return {void}
+     */
+    getComment = (
```
Comment (jstoffan) [inline src/elements/content-sidebar/ActivitySidebar.js:1011]: Minor, but is this needed? Can we call `fetchThreadedComment` directly, instead?

File: src/elements/content-sidebar/activity-feed/activity-feed/ActivityFeed.js [hunk 20]
```diff
@@ -127,6 +140,20 @@ class ActivityFeed extends React.Component<Props, State> {
...
+            onShowReplies &&
+            activeFeedEntryId &&
+            activeFeedEntryType &&
+            (activeFeedEntryType === FEED_ITEM_TYPE_COMMENT || activeFeedEntryType === FEED_ITEM_TYPE_ANNOTATION) &&
+            hasLoaded &&
+            activeFeedEntryReplyId &&
+            currFeedItems &&
+            !isItemInFeed(currFeedItems, activeFeedEntryReplyId)
+        ) {
+            onShowReplies(activeFeedEntryId, activeFeedEntryType);
```
Comment (jstoffan) [inline src/elements/content-sidebar/activity-feed/activity-feed/ActivityFeed.js:155]: This seems very indirect. Can we load the replies if we detect that they're needed within the `fetchFeedItems` chain?

File: src/elements/content-sidebar/ActivitySidebar.js (hunk 14 above)
Comment (kkuliczkowski-box) [reply src/elements/content-sidebar/ActivitySidebar.js:170]: I understand your point but in this case I don't see any alternative really. The `activeFeedEntryId` and `activeFeedEntryType` props represent the values of url params. Previously they were simply passed to `ActivityFeed`, where logic for displaying "selected state" lives.

Now with the possibility of a reply id being set as `activeFeedEntryId` we need to check if it is somewhere within feed items. If it turns out that it's a reply we need to pass its parent as `activeFeedEntryId` as the parent needs to be "selected/highlighted". In that case I'm updating values in state and pass them to `ActivityFeed`. If no action is taken. Moreover if the reply id is not found within feed items, we need to fetch for that reply's data in order to find its parent and then set its id in state. 

Ultimately this is a temporary solution until we get the `partentId` as an url param. With that we would not need to store the active item id in state as it would simply be either the `itemId` or the `parentId`.

File: src/elements/content-sidebar/ActivitySidebar.js (hunk 16 above)
Comment (kkuliczkowski-box) [reply src/elements/content-sidebar/ActivitySidebar.js]: These additional asynchronous actions (fetching reply data and its parent replies) will take place only once when feed items are not yet loaded. From what I can tell, during that time fetchFeedItems will not be called as the whole Activity Sidebar is in the "loading state".

I've also moved that additional actions logic outside of `fetchFeedItemsSuccessCallback`.

File: src/elements/content-sidebar/activity-feed/activity-feed/ActivityFeed.js (hunk 20 above)
Comment (kkuliczkowski-box) [reply src/elements/content-sidebar/activity-feed/activity-feed/ActivityFeed.js]: Yes, that was a wrong approach. I've reverted the `ActivityFeed` completely and fetching for replies is now a part of the `fetchFeedItems` chain during the first load of `ActivitySidebar`.

File: src/elements/content-sidebar/ActivitySidebar.js [hunk 21]
```diff
@@ -530,11 +536,32 @@ class ActivitySidebar extends React.PureComponent<Props, State> {
...
 
+        const feedAPI = api.getFeedAPI(false);
         feedAPI.file = file;
+
+        // Detect if replies are being hidden and activeFeedEntryId belongs to a reply
+        // that is in currently being updated parent, in order to disable active item
+        if (
+            activeFeedEntryId &&
+            replies.length === 1 &&
+            feedItems.find(
```
Comment (jstoffan) [inline src/elements/content-sidebar/ActivitySidebar.js:554]: Should this be `.some`? It looks we do a form of this check in quite a few places. Would it be meaningful to move it to a utility function?

File: src/elements/content-sidebar/ActivitySidebar.js [hunk 22]
```diff
@@ -530,11 +536,32 @@ class ActivitySidebar extends React.PureComponent<Props, State> {
...
+        }
 
+        const feedAPI = api.getFeedAPI(false);
         feedAPI.file = file;
+
+        // Detect if replies are being hidden and activeFeedEntryId belongs to a reply
+        // that is in currently being updated parent, in order to disable active item
+        if (
+            activeFeedEntryId &&
+            replies.length === 1 &&
```
Comment (jstoffan) [inline src/elements/content-sidebar/ActivitySidebar.js:553]: Would this cause any issues with threads that are exactly one reply long?

File: src/elements/content-sidebar/ActivitySidebar.js [hunk 23]
```diff
@@ -671,26 +698,68 @@ class ActivitySidebar extends React.PureComponent<Props, State> {
...
+                    !this.isItemTypeComment(type)
+                ) {
+                    this.fetchFeedItemsSuccessCallback(feedItems);
+                    return;
+                }
+
+                const parentType: CommentFeedItemType =
+                    type === FEED_ITEM_TYPE_COMMENT ? FEED_ITEM_TYPE_COMMENT : FEED_ITEM_TYPE_ANNOTATION;
+
+                this.getFeedItemsWithReplies(feedItems, id, parentType)
```
Comment (jstoffan) [inline src/elements/content-sidebar/ActivitySidebar.js:754]: Can we flatten out this promise to avoid needing the duplicate `.catch`?

```
this.getActiveFeedEntryData(feedItems)
    .then(({ id, type }) => {
        if (
            !id ||
            !type ||
            this.isActiveEntryInFeed(feedItems, activeFeedEntryId) ||
            !this.isItemTypeComment(type)
        ) {
            return Promise.resolve(feedItems);
        }

        const parentType: CommentFeedItemType =
            type === FEED_ITEM_TYPE_COMMENT ? FEED_ITEM_TYPE_COMMENT : FEED_ITEM_TYPE_ANNOTATION;

        return this.getFeedItemsWithReplies(feedItems, id, parentType)
    })
    .then(updatedItems => this.fetchFeedItemsSuccessCallback(updatedItems))
    .catch(error => this.fetchFeedItemsErrorCallback(feedItems, [error]));

**Review by jstoffan — APPROVED** [review]
LGTM as a temporary solution. We'll want to revisit this code in the near future, though.

**Review by wpiesiak — APPROVED** [review]

File: src/elements/content-sidebar/ActivitySidebar.js (hunk 21 above)
Comment (kkuliczkowski-box) [reply src/elements/content-sidebar/ActivitySidebar.js:554]: Right, I've changed to `some` and also noticed that I could actually use a method (`getCommentFeedItemByReplyId`) I've created earlier, that already has logic to find the parent based on reply id and have used it now.
I've merged this PR because of time constraints but opened a new one for these minor fixes: https://github.com/box/box-ui-elements/pull/3148

File: src/elements/content-sidebar/ActivitySidebar.js (hunk 22 above)
Comment (kkuliczkowski-box) [reply src/elements/content-sidebar/ActivitySidebar.js:553]: It should not as threads can be "hidden" only if there are 2 or more replies. Hiding replies basically means setting replies array to just one entry (the last reply).

**Comment by kkuliczkowski-box** [comment]
I've created a new PR in order to address the latest comments: https://github.com/box/box-ui-elements/pull/3148
